### PR TITLE
Contributing: Introduce some general commit rules

### PR DESCRIPTION
**Pull request by nanoant**
Due some latest changes that introduced some substantial noise in the project
history I hereby propose some general commit rules that have proven to work in
many other projects.

Both contributors and maintainers (accepting PRs) are obliged to follow this
rules, for sake of the Nim project and good collaboration.

### CODE REVIEW

File: contributing.rst [hunk 1]
```diff
@@ -1,7 +1,46 @@
...
+
+   Always check your changes for whitespace errors using ``git diff --check`` or add
+   following ``pre-commit`` hook:
+
+   .. code-block:: sh
+
+      #!/bin/sh
+      git diff --check --cached || exit $?
+
+   Nim does not care about trailing whitespace, neither reStructuredText, so you should
```
Comment (Araq) [inline contributing.rst]: Make that: "No sane programming or markup language cares about trailing whitespace." :P

**Comment by reactormonk** [comment]
Let's just dunk this problem via https://github.com/nim-lang/Nim/issues/2967

**Comment by nanoant** [comment]
@Araq Edited the whitespace reasoning to mention language sanity :smiling_imp:

File: contributing.rst [hunk 2]
```diff
@@ -1,7 +1,49 @@
-The git stuff
+The Git stuff
 =============
 
-`Guide by github, scroll down a bit <https://guides.github.com/activities/contributing-to-open-source/>`_
+General commit rules
+--------------------
+
+1. All changes introduced by the commit (diff lines) must be related to the
```
Comment (reactormonk) [inline contributing.rst]: Change to commits. I prefer more commits over squashing. Squashing makes it hard to follow the PR discussions.

File: contributing.rst [hunk 3]
```diff
@@ -1,7 +1,49 @@
...
+   subject of the commit.
+
+   If you change some other unrelated to the subject parts of the file, because
+   your editor reformatted automatically the code or whatever different reason,
+   this should be excluded from the commit.
+
+   *Tip:* Never commit everything as is using ``git commit -a``, but review
+   carefully your changes with ``git add -p``.
+
+2. Changes should not introduce any trailing whitespace.
```
Comment (reactormonk) [inline contributing.rst]: I'd just delete this part and solve it via https://github.com/nim-lang/Nim/issues/2967 - via gitattributes.

Comment (nanoant) [inline contributing.rst]: Git attributes AFAIK solve line-endings problem, but not trailing whitespace. We may advertise to trim it via some favorite editor settings.

Comment (reactormonk) [inline contributing.rst]: imo getting rid of your own trailing white space should be common sense - should we write that part down?

File: contributing.rst (hunk 2 above)
Comment (nanoant) [inline contributing.rst]: @reactormonk I am sorry, but I don't understand your suggestion, I should change what to "commits"? I am not talking about PR, but rules for single commit, the subject is first line of the commit description.

Comment (reactormonk) [inline contributing.rst]: Oh yeah, sorry. Sounds good to me.

**Comment by reactormonk** [comment]
Could we add a passage against squashing & force-pushing? It makes PR discussions harder to follow. I assume the trade-off is a bit more complicated history, but `./koch` should be able to deal with incorrect code when bisecting.

File: contributing.rst (hunk 3 above)
Comment (reactormonk) [inline contributing.rst]: The less text we have, the more likely people will actually read it.

Comment (nanoant) [inline contributing.rst]: > imo getting rid of your own trailing white space should be common sense - should we write that part down?

I wouldn't bother, but for some reason that enormous number of trailing whitespace landed to the Nim repository :wink:

Comment (reactormonk) [inline contributing.rst]: I'm sure the responsible parties have learned and meanwhile posted a few rants about git not ignoring whitespace changes.

File: contributing.rst [hunk 4]
```diff
@@ -1,7 +1,49 @@
...
+      git diff --check --cached || exit $?
+
+   No sane programming or markup language cares about trailing whitespace, so
+   tailing whitespace is just a noise you should not introduce to the
+   repository.
+
+3. Describe your commit well following the 50/80 rule on commit messages:
+
+   Start with the commit subject as single line maximum of 50 characters,
+   without trailing dot, briefly describing the change.
```
Comment (dom96) [inline contributing.rst]: Why "without trailing dot"?

**Comment by Araq** [comment]
One reason for the trailing whitespace in the Nim repository is that it's simply much easier to generate and most Nim code in the compiler has been generated (translated from Pascal).

And it never occured to me that 9 years later the world of software development would have gone fully insane with ever more bullshit bingo rules that have nothing to do with software quality, scientific rigor or plain and simple common sense.

So thanks, now I need to care about Unix executable bits (a broken piece of junk, just look at the file, if it's ELF or starts with #!, it's executable; or even better: I like to execute it if it's the first whitespace separated thing in the command...), using the right (TM) line endings, not producing trailing whitespace but producing files which *end* in a line ending (which for some reason doesn't count as trailing whitespace...), not using the UTF-8 BOM (even though in other cases looking at file name extensions to determine the file type is evil and we should instead look at the file's magic byte headers), splitting my trivial changes in trivial separate commit messages just because somebody perhaps likes to masturbate over git's "History" while looking at in a 13 baud 80x20 character console.

When working at Nim's core feels more tedious than the job that I actually get paid for, you can be sure we're on the right track and everything is smooth and "professional".

**Comment by Araq** [comment]
So ... end of rant. Please note that I didn't close the PR. ;-)

**Comment by reactormonk** [comment]
A note on the BOM: It didn't get adopted because of `#!`. It breaks shebangs.

File: contributing.rst (hunk 4 above)
Comment (nanoant) [inline contributing.rst]: > Why "without trailing dot"?

Because this is what 50/80 rule is, see: A Note About Git Commit Messages link below for more explanation.

Comment (dom96) [inline contributing.rst]: That link doesn't seem to mention anything about the trailing dot.

**Comment by nanoant** [comment]
@Araq Hahaha... good one. Yeah I see your point. But I can tell you, you got what you wanted:

1. You made this project compile on UNIX (OS X, Linux), that attracted some UNIX junkies like myself that believe some stupid execution bit is important.

2. You published your project on GitHub, that made people read your commit logs, than may be not really sexually appealing (unless you embed some nude pics urls there), but are just interesting source of knowledge.

Now, you have to decide if you care if some people will run Nim on UNIX and read your Git logs, comment, post stupid ideas, propose to introduce `end` keyword, claim on your forums that UNIX rulez, while Windows sucks llama's ass (even you think opposite). It is all yours, you have unleashed the hell of the open-source project and like it or now, you've created own small community.

P.S. I don't mind if you close the PR, I don't mind if don't care about whitespace or fairy dust, I don't mind if you ask me to shut up. I am just all happy that GitHub lets me express my opinion and also enjoy reading other's (often opposite) opinions on the subject too :)

File: contributing.rst (hunk 4 above)
Comment (nanoant) [inline contributing.rst]: @dom96 You are right, it does not explicitly mention trailing period (I should call it "period" probably). Here is another one: http://chris.beams.io/posts/git-commit/

Comment (dom96) [inline contributing.rst]: Funny that this came up on HN http://antirez.com/news/90

Comment (nanoant) [inline contributing.rst]: @dom96 Here a good quote from the comments of the article you mentioned:

antirez writes:
> When you're a part of a team (whether work or whatever) and there is no reason to do things either way (putting a dot vs not putting a dot or tabs vs spaces), JUST FOLLOW THE CONVENTION. Stop being a drama queen.

Comment (dom96) [inline contributing.rst]: Yes, the convention has never been "No trailing dot". This does not give any reason as to why this convention is necessary. I will remove this part from the guide unless you give me a good reason as to the rationale.

Comment (reactormonk) [inline contributing.rst]: Ok, then make it "use a line you would also use as an email topic".

Comment (nanoant) [inline contributing.rst]: @dom96 You are now kidding, don't you? You are pasting a link to the article that states: *"Nor subjects, for what matters. Everybody will tell you to don't add a dot at the end of the first line of a commit message."* and now you are claiming that that's not a convention. Let me quote Google:

> Convention:
> a way in which something is usually done, especially within a particular area or activity.

But yes whatever, you have free will, do what you want.

Comment (nanoant) [inline contributing.rst]: > Ok, then make it "use a line you would also use as an email topic".

@reactormonk Very reasonable :+1:

Comment (dom96) [inline contributing.rst]: It's not a convention for *this project*. Should we all just blindly follow what everybody else is doing without knowing *why* they are doing it?

I don't care if you use a trailing dot or not, why do you want to restrict me from using one if I so choose?
